### PR TITLE
Add support for defining credentials for uv in an environment variable

### PR DESCRIPTION
**Pull request by zanieb**
Adds `UV_AUTH_URLS` which defines credentials that should be used for URLs if we make a request there.

This is helpful for cases where, e.g., you want to provide credentials for an index but you don't want to change the index URL semantics by setting `UV_INDEX_URL`.

Implemented by seeding the URL authentication cache with the value of the variable.

Supports things like:

- `UV_AUTH_URLS="username:password@hostname"`
- `UV_AUTH_URLS="username@hostname"`
- `UV_AUTH_URLS="username:password@hostname username:password@other-hostname"`
- `UV_AUTH_URLS="username:password@hostname/prefix"`

There's a small caveat here, that if you use an index URL like `foo@hostname/simple` and you provide a password `bar:password@hostname` — the password should not be used but requests to fetch wheels from a different, parent path e.g.`hostname/files` can allow use of `bar:password`. This requires #4583 to address. In the meantime, credentials should just be properly scoped.

### CODE REVIEW

File: crates/uv-auth/src/lib.rs
Comment (zanieb) [inline crates/uv-auth/src/lib.rs:79]: @BurntSushi recommendations here?

Comment (zanieb) [reply crates/uv-auth/src/lib.rs:79]: (We can't try parsing and reparse on a missing scheme error because `public:test@foo.com` treats `public` as a scheme.)

Comment (BurntSushi) [reply crates/uv-auth/src/lib.rs:79]: What's wrong with what's here?

Comment (zanieb) [reply crates/uv-auth/src/lib.rs:79]: Uhh, well if you do `http://foo` we don't want to support that but we'll fail by doing `https://http://foo` which seems weird. I guess that's fine though?

Comment (zanieb) [reply crates/uv-auth/src/lib.rs]: I'll just special case http for now.

**Comment by zanieb** [comment]
I want to write documentation for this and make sure it's actually solving people's problems, but the implementation is ready for review.

File: crates/uv-auth/src/lib.rs
Comment (BurntSushi) [reply crates/uv-auth/src/lib.rs]: Yeah I don't have any better ideas here, especially if URL parsing is treating things as schemes that we don't want to be treated as schemes. So I'd guess you'd probably need to write a custom URL parser to deal with this in a way that is completely satisfying. Special casing `http` seems fine.

File: crates/uv-auth/src/lib.rs
Comment (BurntSushi) [inline crates/uv-auth/src/lib.rs:53]:
```suggestion
/// added to the realm-level cache to ensure that a URL with a path is not used unless a request
```

File: crates/uv-auth/src/lib.rs
Comment (BurntSushi) [inline crates/uv-auth/src/lib.rs:87]: Should this be a `warn!`?

**Review by BurntSushi — APPROVED** [review]

File: crates/uv-auth/src/lib.rs
Comment (zanieb) [reply crates/uv-auth/src/lib.rs:87]: Can't use our `warn_user` macros here, but we could use a warn tracing level — just doesn't mean much yet. Seems reasonable though.

Comment (zanieb) [reply crates/uv-auth/src/lib.rs:87]:
```suggestion
        warn!("Ignoring insecure URL in `UV_AUTH_URLS`: {url}");
```

File: crates/uv-auth/src/lib.rs
Comment (zanieb) [inline crates/uv-auth/src/lib.rs:3]:
```suggestion
use tracing::{debug, trace, warn};
```

File: crates/uv-auth/src/lib.rs
Comment (zanieb) [inline crates/uv-auth/src/lib.rs:75]:
```suggestion
            warn!("Ignoring URL without credentials in `UV_AUTH_URLS`: {url}");
```

**Comment by charliermarsh** [comment]
Maybe `UV_BASIC_AUTH_URLS`? Feel free to ignore though.

**Comment by zanieb** [comment]
I don't love the name, so maybe. Will consider...

File: crates/uv/tests/pip_install.rs
Comment (mkniewallner) [inline crates/uv/tests/pip_install.rs:3857]:
```suggestion
fn install_package_basic_auth_from_uv_basic_auth_urls() {
```
grep and replace gone a bit too wild I guess 😄

**Comment by mkniewallner** [comment]
Since auto-merge was enabled, apart from test and `clippy` failures, is there anything preventing the merge of this PR, or was it maybe just forgotten?

File: crates/uv/tests/pip_install.rs
Comment (zanieb) [reply crates/uv/tests/pip_install.rs:3857]: Haha thanks

**Comment by zanieb** [comment]
I'm not getting a ton of signal about how useful it would be so it's relatively low priority.

More relevant though, I was on vacation and am just how noticing auto-merge failed.

**Comment by charliermarsh** [comment]
@zanieb -- My gut is to punt on this until we see more demand, but ultimately defer to you.

**Comment by mkniewallner** [comment]
> I'm not getting a ton of signal about how useful it would be so it's relatively low priority.

Definitely not high priority IMO. Personally I can leave without it, since we can pass the entire index including credentials with `UV_INDEX_URL` and `UV_EXTRA_INDEX_URL`. So it's really just like a convenient thing to have, as I like the idea of having the index explicitly set in `[tool.uv]`. On projects where a lot of developers work on, having this explicit configuration makes things more obvious (although in fact, even without the feature, you can still explicitly set the index in `[tool.uv]` and override the entire thing with the env var, but this kinda repeats something that is already set).

Do you think the feature would also fit what is planned in https://github.com/astral-sh/uv/issues/171, or are we thinking about providing another mechanism to provide credentials for this other feature (e.g. what [Poetry does](https://python-poetry.org/docs/repositories/#configuring-credentials), where the credentials are set for a source name by including the name in the header)? If we're not sure yet, maybe we can hold off on merging this PR until the other feature is implemented, to try to come up with something that would ideally fit both use cases?

**Comment by zanieb** [comment]
@charliermarsh is going to start working on #171 soon and I think that will probably supersede this and include a design for index credentials.

**Pull request closed without merging**
